spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -17,80 +17,352 @@
-        // Calculate the number of files missing `numRecords`
-        count(when(col("stats.numRecords").isNull, 1)))
-      .first
+    val filesWithStatsForScan = deltaScanGenerator.filesWithStatsForScan(Nil)
+    // Validate all the files has stats
+    val filesStatsCount = filesWithStatsForScan.select(
+      sum(numLogicalRecords).as("numLogicalRecords"),
+      count(when(col("stats.numRecords").isNull, 1)).as("missingNumRecords"),
+      count(when(col("stats.numRecords") > 0, 1)).as("countNonEmptyFiles")).head
 
-    // The count agg is never null. A non-zero value means we have incomplete stats; otherwise,
+    // If any numRecords is null, we have incomplete stats;
+    val allRecordsHasStats = filesStatsCount.getAs[Long]("missingNumRecords") == 0
+    if (!allRecordsHasStats) {
+      return (None, Map.empty)
+    }
     // the sum agg is either null (for an empty table) or gives an accurate record count.
-    if (row.getLong(1) > 0) return None
-    val numRecords = if (row.isNullAt(0)) 0 else row.getLong(0)
-    Some(numRecords)
+    val numRecords = if (filesStatsCount.isNullAt(0)) 0 else filesStatsCount.getLong(0)
+    lazy val numFiles: Long = filesStatsCount.getAs[Long]("countNonEmptyFiles")
+
+    val dataColumns = snapshot.statCollectionPhysicalSchema.filter(col =>
+      AggregateDeltaTable.isSupportedDataType(col.dataType) &&
+      lowerCaseColumnNames.contains(col.name.toLowerCase(Locale.ROOT)))
+
+    // DELETE operations creates AddFile records with 0 rows, and no column stats.
+    // We can safely ignore it since there is no data.
+    lazy val files = filesWithStatsForScan.filter(col("stats.numRecords") > 0)
+    lazy val statsMinMaxNullColumns = files.select(col("stats.*"))
+    if (dataColumns.isEmpty
+      || !isTableDVFree(snapshot)
+      || numFiles == 0
+      || !statsMinMaxNullColumns.columns.contains("minValues")
+      || !statsMinMaxNullColumns.columns.contains("maxValues")
+      || !statsMinMaxNullColumns.columns.contains("nullCount")) {
+      return (Some(numRecords), Map.empty)
+    }
+
+    // dataColumns can contain columns without stats if dataSkippingNumIndexedCols
+    // has been increased
+    val columnsWithStats = files.select(
+      col("stats.minValues.*"),
+      col("stats.maxValues.*"),
+      col("stats.nullCount.*"))
+      .columns.groupBy(identity).mapValues(_.size)
+      .filter(x => x._2 == 3) // 3: minValues, maxValues, nullCount
+      .map(x => x._1).toSet
+
+    // Creates a tuple with physical name to avoid recalculating it multiple times
+    val dataColumnsWithStats = dataColumns.map(x => (x, DeltaColumnMapping.getPhysicalName(x)))
+      .filter(x => columnsWithStats.contains(x._2))
+
+    val columnsToQuery = dataColumnsWithStats.flatMap { columnAndPhysicalName =>
+      val dataType = columnAndPhysicalName._1.dataType
+      val physicalName = columnAndPhysicalName._2
+
+      Seq(col(s"stats.minValues.`$physicalName`").cast(dataType).as(s"min.$physicalName"),
```

Comment:
nit: make minValues/maxValues/nullCount const string and reuse all across.

Reply:
Done